### ISSUE TITLE
terrible torch.nn.embeddingbag performance on TPU

### ISSUE THREAD

**Issue by shz0116** [pytorch/xla#2318]
I am measuring the performance of torch.nn.embeddingbag on TPU using colab (google notebook).
I found the performance is terrible, 100x-1000x slow than CPU and GPU.
The algo mem bw I got on CPU and GPU are 42 GB/s and 400 GB/s.
Here is only 0.06 GB/s.
 I am wondering where I did wrong.
Here is the code I used for the test and output with performance metric:

```python
print(dev)
import time
import torch_xla.debug.metrics as met 

num_features=3000000
batch_size=8192
nnz = 100 
embed_dim = 128 
time1 = time.perf_counter()
h_indices  = torch.randint(0, num_features, (batch_size, nnz)).to(dev)
time2 = time.perf_counter()
print(h_indices.device)
print("Creating indices take ", time2 - time1, " seconds")
time3 = time.perf_counter()
h_emb = torch.nn.EmbeddingBag(num_features, embed_dim, mode='sum').to(dev)
time4 = time.perf_counter()
print("Creating emb take ", time4 - time3, " seconds")
h_results = torch.zeros(batch_size, embed_dim)
total_bytes = batch_size * nnz * embed_dim * h_emb.weight.element_size()
print("TPU: total bytes: ", total_bytes)

for i in range(4):
  time5 = time.perf_counter()
  h_results = h_emb(h_indices)
  time6 = time.perf_counter()
  torch_xla._XLAC._xla_sync_multi([h_results], devices=[], wait=True, sync_xla_data=True)
  time7 = time.perf_counter()
  print("EMB take {0:0.6f} sec with sync {1:.6f}".format(time6 - time5, time7- time5))
  print("TPU: algo mem bw ", total_bytes*1.0/(time7- time5)/1.0e9, "GB/s")

h_results = h_results.cpu()
print(met.metrics_report())
print(h_results)
print(h_results.device)
```

Output
========
xla:1
xla:1
Creating indices take  0.15045019499939372  seconds
Creating emb take  20.371595505000187  seconds
TPU: total bytes:  419430400
EMB take 6.780893 sec with sync 6.825369
TPU: algo mem bw  0.06145168329124725 GB/s
EMB take 6.747784 sec with sync 6.792259
TPU: algo mem bw  0.06175124148685416 GB/s
EMB take 8.182684 sec with sync 8.227786
TPU: algo mem bw  0.05097730861427444 GB/s
EMB take 7.377474 sec with sync 7.421842
TPU: algo mem bw  0.05651297980530163 GB/s
Metric: CompileTime
  TotalSamples: 29
  Accumulator: 741ms839.943us
  ValueRate: 003ms706.047us / second
  Rate: 0.105928 / second
  Percentiles: 1%=007ms140.722us; 5%=007ms150.140us; 10%=007ms328.325us; 20%=008ms526.751us; 50%=009ms660.796us; 80%=019ms656.743us; 90%=082ms935.935us; 95%=087ms195.651us; 99%=259ms084.098us
Metric: DeviceLockWait
  TotalSamples: 46
  Accumulator: 117.399us
  ValueRate: 000.286us / second
  Rate: 0.112224 / second
  Percentiles: 1%=001.409us; 5%=001.464us; 10%=001.500us; 20%=001.643us; 50%=002.165us; 80%=003.333us; 90%=004.030us; 95%=005.039us; 99%=005.390us
Metric: ExecuteTime
  TotalSamples: 36
  Accumulator: 250ms480.704us
  ValueRate: 622.480us / second
  Rate: 0.0894651 / second
  Percentiles: 1%=002ms993.405us; 5%=002ms138.450us; 10%=002ms232.198us; 20%=002ms411.969us; 50%=003ms694.508us; 80%=016ms950.929us; 90%=016ms361.433us; 95%=017ms688.248us; 99%=017ms756.986us
Metric: InboundData
  TotalSamples: 38
  Accumulator: 11.49GB
  ValueRate: 28.72MB / second
  Rate: 0.0927306 / second
  Percentiles: 1%=0.00B; 5%=0.00B; 10%=1.00B; 20%=1.00B; 50%=4.00B; 80%=1.44GB; 90%=1.44GB; 95%=1.44GB; 99%=1.44GB
Metric: IrValueTensorToXlaData
  TotalSamples: 5
  Accumulator: 37s193ms387.850us
  ValueRate: 098ms748.073us / second
  Rate: 0.0131405 / second
  Percentiles: 1%=004ms214.232us; 5%=004ms214.232us; 10%=004ms214.232us; 20%=138ms990.848us; 50%=162ms241.477us; 80%=19s355ms156.085us; 90%=19s355ms156.085us; 95%=19s355ms156.085us; 99%=19s355ms156.085us
Metric: OutboundData
  TotalSamples: 17
  Accumulator: 2.90GB
  ValueRate: 7.26MB / second
  Rate: 0.0414777 / second
  Percentiles: 1%=4.00B; 5%=4.00B; 10%=4.00B; 20%=4.00B; 50%=4.00MB; 80%=6.25MB; 90%=1.43GB; 95%=1.43GB; 99%=1.43GB
Metric: ReleaseDataHandlesTime
  TotalSamples: 45
  Accumulator: 130ms101.501us
  ValueRate: 317.523us / second
  Rate: 0.109826 / second
  Percentiles: 1%=592.856us; 5%=628.579us; 10%=654.615us; 20%=804.820us; 50%=001ms055.432us; 80%=002ms884.291us; 90%=010ms630.161us; 95%=014ms248.080us; 99%=016ms284.852us
Metric: TensorsGraphSize
  TotalSamples: 36
  Accumulator: 228.00
  ValueRate: 0.57 / second
  Rate: 0.0894666 / second
  Percentiles: 1%=3.00; 5%=3.00; 10%=3.00; 20%=3.00; 50%=5.00; 80%=12.00; 90%=12.00; 95%=12.00; 99%=12.00
Metric: TransferFromServerTime
  TotalSamples: 38
  Accumulator: 52s525ms158.777us
  ValueRate: 126ms735.724us / second
  Rate: 0.0927306 / second
  Percentiles: 1%=002.465us; 5%=002.599us; 10%=001ms212.485us; 20%=001ms260.495us; 50%=002ms730.394us; 80%=06s563ms950.798us; 90%=06s224ms065.128us; 95%=07s985ms079.807us; 99%=08s652ms129.998us
Metric: TransferToServerTime
  TotalSamples: 17
  Accumulator: 38s567ms142.087us
  ValueRate: 092ms650.793us / second
  Rate: 0.0414741 / second
  Percentiles: 1%=001ms270.987us; 5%=001ms270.987us; 10%=001ms279.093us; 20%=006ms534.591us; 50%=044ms239.890us; 80%=138ms972.689us; 90%=18s534ms720.856us; 95%=19s355ms117.506us; 99%=19s355ms117.506us
Metric: TransferToServerTransformTime
  TotalSamples: 17
  Accumulator: 02s923ms428.756us
  ValueRate: 005ms692.907us / second
  Rate: 0.0414777 / second
  Percentiles: 1%=148.738us; 5%=148.738us; 10%=162.302us; 20%=305.740us; 50%=001ms234.812us; 80%=002ms853.995us; 90%=473ms088.622us; 95%=01s433ms805.262us; 99%=01s433ms805.262us
Counter: CachedCompile
  Value: 7
Counter: CreateCompileHandles
  Value: 29
Counter: CreateDataHandles
  Value: 71
Counter: CreateXlaTensor
  Value: 131
Counter: DestroyDataHandles
  Value: 64
Counter: DestroyXlaTensor
  Value: 124
Counter: ReleaseDataHandles
  Value: 65
Counter: SyncTensorsToData
  Value: 8
Counter: UncachedCompile
  Value: 29
Counter: XRTAllocateFromTensor_Empty
  Value: 6
Counter: XrtCompile_Empty
  Value: 384
Counter: XrtExecuteChained_Empty
  Value: 384
Counter: XrtExecute_Empty
  Value: 384
Counter: XrtRead_Empty
  Value: 384
Counter: XrtReleaseAllocationHandle_Empty
  Value: 384
Counter: XrtReleaseCompileHandle_Empty
  Value: 384
Counter: XrtSessionCount
  Value: 4
Counter: XrtSubTuple_Empty
  Value: 384
Counter: aten::_embedding_bag
  Value: 8
Counter: aten::_local_scalar_dense
  Value: 20
Counter: aten::masked_select
  Value: 2
Counter: xla::abs
  Value: 3
Counter: xla::arange_out
  Value: 8
Counter: xla::as_strided
  Value: 4
Counter: xla::bitwise_and_out
  Value: 2
Counter: xla::ceil
  Value: 9
Counter: xla::copy_
  Value: 13
Counter: xla::div
  Value: 1
Counter: xla::empty
  Value: 17
Counter: xla::empty_strided
  Value: 4
Counter: xla::eq
  Value: 2
Counter: xla::fill_
  Value: 1
Counter: xla::gt
  Value: 2
Counter: xla::masked_select
  Value: 2
Counter: xla::max
  Value: 1
Counter: xla::min
  Value: 1
Counter: xla::mul
  Value: 2
Counter: xla::ne
  Value: 13
Counter: xla::select
  Value: 24
Counter: xla::view
  Value: 10
Counter: xla::zero_
  Value: 1
Metric: XrtAllocateFromTensor
  TotalSamples: 17
  Accumulator: 02s341ms083.680us
  Mean: 138ms710.805us
  StdDev: 346ms041.949us
  Rate: 0.0414752 / second
  Percentiles: 25%=887.137us; 50%=002ms276.757us; 80%=070ms891.900us; 90%=01s049ms291.705us; 95%=01s116ms869.379us; 99%=01s116ms869.379us
Metric: XrtCompile
  TotalSamples: 29
  Accumulator: 469ms132.671us
  Mean: 016ms176.989us
  StdDev: 045ms498.104us
  Rate: 0.105931 / second
  Percentiles: 25%=006ms488.203us; 50%=007ms054.814us; 80%=009ms769.816us; 90%=010ms453.599us; 95%=014ms979.407us; 99%=257ms780.586us
Metric: XrtExecute
  TotalSamples: 36
  Accumulator: 155ms728.861us
  Mean: 004ms298.024us
  StdDev: 006ms569.987us
  Rate: 0.0894673 / second
  Percentiles: 25%=001ms202.144us; 50%=001ms383.589us; 80%=014ms154.679us; 90%=015ms696.424us; 95%=015ms201.621us; 99%=015ms205.583us
Metric: XrtExecutorEvict
  TotalSamples: 0
  Accumulator: nanB
  Mean: nanB
  StdDev: nanB
  Percentiles: 
Metric: XrtReadLiteral
  TotalSamples: 54
  Accumulator: 29s303ms706.294us
  Mean: 543ms642.709us
  StdDev: 964ms018.653us
  Rate: 0.133263 / second
  Percentiles: 25%=501.071us; 50%=741.740us; 80%=633ms675.320us; 90%=03s645ms631.217us; 95%=03s837ms289.574us; 99%=03s473ms278.951us
Metric: XrtReleaseAllocation
  TotalSamples: 46
  Accumulator: 001ms431.450us
  Mean: 031.118us
  StdDev: 027.594us
  Rate: 0.112257 / second
  Percentiles: 25%=015.575us; 50%=020.295us; 80%=038.175us; 90%=089.258us; 95%=095.506us; 99%=106.722us

tensor([[  7.2877,   5.1902,   1.9751,  ...,  17.8653,   5.0835,   3.2784],
        [ -6.0403,  11.0338, -16.1487,  ...,  -6.9018, -17.2799,  13.7087],
        [-16.4450,  14.2040, -14.2556,  ...,  24.1469, -24.0838, -12.1110],
        ...,
        [ -2.5326, -17.8629,   5.2641,  ...,  20.6966,   7.3132,   6.2163],
        [ 33.5090,   0.5714,   3.5419,  ...,  20.3194,  24.2440,  19.0229],
        [ -6.4958,  -6.2275,  13.1636,  ...,  -4.6331,  18.9097,   3.6664]],
       grad_fn=<CopyBackwards>)
cpu

**Comment by JackCaoG** [comment]
Embeedingbag lowering is not implemented on the PT/XLA so it is forwarded back to CPU which is what caused the slowness. Note that 
```
Counter: aten::embedding_bag
Value: 8
Counter: aten::local_scalar_dense
Value: 20
Counter: aten::masked_select
Value: 2
```

All of the aten counters represent the call forwarded to pytorch CPU. Our support for masked_select is experimental(forwarded to CPU by default), since it involves dynamic shape. @taylanbil I remember you work around embeddingBag by a combination of other ops. Maybe that will be helpful?

**Comment by taylanbil** [comment]
this is what I'm using temporarily to sidestep the usage of `nn.EmbeddingBag`: https://github.com/taylanbil/dlrm/blob/tpu/tools/xla_embedding_bag.py

@JackCaoG is the PT change required for backward done? Are we blocked on anything to implement this on xla?

**Comment by JackCaoG** [comment]
I think pytorch change merged last week, we should be good on that end.

**Comment by ailzhang** [comment]
Pytorch change is not yet merged but should be soon I think. cc: @gmagogsfm 
Once the pytorch change is merged, a lowering in pytorch/xla is still needed I think.

**Comment by ailzhang** [comment]
FYI the Pytorch change has been merged. Thanks @gmagogsfm !
